docs/resource/README.md

```diff
@@ -102,7 +102,104 @@ as specified in the [Resource SDK specification](https://github.com/open-telemet
+**[1]:** MUST be unique for each instance of the same `service.namespace,service.name` pair
+(in other words `service.namespace,service.name,service.instance.id` triplet MUST be globally unique).
+The ID helps to distinguish instances of the same service that exist at the same time
+(e.g. instances of a horizontally scaled service). It is preferable for the ID to be persistent
+and stay the same for the lifetime of the service instance, however it is acceptable that
+the ID is ephemeral and changes during important lifetime events for the service
+(e.g. service instance restarts).
+
+If the instance has no inherent unique ID that can be used as the value of this attribute,
+implementations MAY generate a random Version 1 or Version 4
+[RFC 4122](https://www.ietf.org/rfc/rfc4122.txt) UUID. When enough data is available,
+implementations SHOULD use Version 5 and SHOULD use the following UUID as the namespace:
+`4d63009a-8d0f-11ee-aad7-4c796ed8e320`.
+
+UUIDs are typically recommended, as only an opaque yet reproducible value for
+the purposes of identifying a service instance is needed. Similar to what can be seen in the man page for the
+[`/etc/machine-id`](https://www.freedesktop.org/software/systemd/man/machine-id.html) file,
+the underlying data, such as pod name and namespace should be treated as
+confidential by this algorithm, being the user's choice to expose it or not via another resource attribute.
+
+When a UUID v5 is generated, the input SHOULD be prefixed with
+`${telemetry.sdk.name}.${telemetry.sdk.language}.${service.namespace}.${service.name}`, followed by the
+the workload identifier, which should tentatively be stable. The workload identifier is the final piece of
+information making the source of telemetry to be uniquely idenfiable. For instance, it might be the `container.name`
+in containerized or Kubernetes environments. This means that the same service yields the
+same UUID if the same identifier (`host.id`, `/etc/machine-id`, and so on) remains the same. It would still
+yield different results for different services on the same host or namespace. When no namespaces or equivalent fields
+are available, the prefix MUST then be `${telemetry.sdk.name}.${telemetry.sdk.language}.${service.name}`.
+
+SDKs MUST provide a `service.instance.id`. SDKs SHOULD use the following algorithm when generating `service.instance.id`:
+
+- If the user has provided a `service.instance.id`, via environment
+  variable, configuration or custom resource detection, this MUST take priority over generated IDs.
```

Comment:
A small nit, but I wonder whether it makes sense to move this out of the algorithm description. Here it's defined as part of an algorithm that _SHOULD_ be used, at least theoretically leaving room to override an instance id specified by a user (which should never happen).

In general, I'm not 100% what the implications are of a "MUST" requirement in the description of an algorithm that "SHOULD" be used. I guess it's not technically a "MUST", as you can decide not to use the algorithm? 

For that reason, I'd prefer to make the first point a "MUST" outside the algorithm definition (it applies regardless of what algorithm is used), and formulate the rest of the algorithm without normative language, like:
> 1. If `k8s.namespace.name`, `k8s.pod.name`, and `k8s.container.name` are available, generate a UUID v5 with the input `${telemetry.sdk.name}.${telemetry.sdk.language}.${k8s.namespace.name}.${service.name}.${k8s.pod.name}.${k8s.container.name}`.
> 2. If `container.id` is available, create an UUID v5 with the input `${telemetry.sdk.name}.${telemetry.sdk.language}.${service.namespace}.${service.name}.${container.id}`
> 3. If `container.id` is available, create an UUID v5 with the input `${telemetry.sdk.name}.${telemetry.sdk.language}.${service.namespace}.${service.name}.${host.id}`.
> 4. If `/etc/machine-id` is available (or alternatively `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Cryptography\MachineGuid` on Windows), create an UUID v5 with the input `${telemetry.sdk.name}.${telemetry.sdk.language}.${service.namespace}.${service.name}.${machine.id}`.
> 5. Otherwise, generate a value using UUID v1 or v4.

Reply:
I think this deserves a discussion on the PR that I'll open as part of the the spec. But one part I can address right now:

> In general, I'm not 100% what the implications are of a "MUST" requirement in the description of an algorithm that "SHOULD" be used. I guess it's not technically a "MUST", as you can decide not to use the algorithm?

What I had in mind was: SDKs are likely unaware of service.instance.id right now and might stay like that for the foreseeable future. Some SDKs might decide to implement this algorithm, and if they do so, they should really follow this algorithm.

Of course, we can still discuss whether a MUST is warranted here, so that all SDKs are aligned in how they do things, or if a SHOULD has to be used, so that they may align, but don't have to.